crates/ide/src/inlay_hints.rs

```diff
@@ -193,14 +194,67 @@ fn get_bind_pat_hints(
         return None;
     }
 
-    acc.push(InlayHint {
-        range: pat.syntax().text_range(),
-        kind: InlayKind::TypeHint,
-        label: ty.display_truncated(sema.db, config.max_length).to_string().into(),
-    });
+    let db = sema.db;
+    if let Some(hint) = hint_iterator(sema, config, &ty, pat.clone()) {
+        acc.push(hint);
+    } else {
+        acc.push(InlayHint {
+            range: pat.syntax().text_range(),
+            kind: InlayKind::TypeHint,
+            label: ty.display_truncated(db, config.max_length).to_string().into(),
+        });
+    }
+
     Some(())
 }
 
+/// Checks if the type is an Iterator from std::iter and replaces its hint with an `impl Iterator<Item = Ty>`.
+fn hint_iterator(
+    sema: &Semantics<RootDatabase>,
+    config: &InlayHintsConfig,
+    ty: &Type,
+    pat: ast::IdentPat,
+) -> Option<InlayHint> {
+    let db = sema.db;
+    let strukt = ty.as_adt()?;
+    let krate = strukt.krate(db)?;
+    if krate.declaration_name(db).as_deref() != Some("core") {
+        return None;
+    }
+    // assert this type comes from `core::iter`
+    strukt
+        .module(db)
+        .path_to_root(db)
+        .into_iter()
+        .rev()
+        .find(|module| module.name(db) == Some(known::iter))?;
```

Comment:
The way we use `known::item` here and `core_iter_Iterator()` is interesting.
We seem to use more or less the same semantics here (some predefined knowledge about sdlib), both using `hir` inside, yet in totally different parts of the code.

Any thoughts? (maybe @matklad ?)

Reply:
I suppose we could add module lookup to `FamousDefs` and then check if the module exports the type or something along those lines? Then we wouldn't have to use `known` here I think

Reply:
Yeah, I am not sure what this should look like, but funneling everything via `FamousDefs` seems prudent for future refactorability. 

I think the right worklist here:

* make sure that IDE bits only use FamousDefs
* replace `FamousDefs::FIXTURE` with more fine-grained opt-ins at the fixture level (we want to have syntax like `//- core with:Iterator,IntoIterator`)
* try to replace as much of path lookup as possible with lang-items and diagnostics items, but keep an eye on ide/compiler split

For this PR, only keeping everything scoped to FamousDefs should be sufficient.